singularity-eos/eos/eos_builder.hpp

```diff
@@ -15,129 +15,116 @@
+          typename... Args>
+Variant<Ts...> ModifyHelper(std::true_type, const Variant<Ts...> &var,
+                            const type_list<U, Rest...> &tl, Args &&...args);
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyHelper(std::false_type, const Variant<Ts...> &var,
+                            const type_list<U, Rest...> &tl, Args &&...args);
 
-template <typename T>
-EOS makeRelativistic(T &&eos, Real cl) {
-  return RelativisticEOS<T>(std::move(eos), cl);
+// Base case
+template <template <class...> typename Mod, typename... Ts, typename... Args>
+Variant<Ts...> ModifyDispatcher(const Variant<Ts...> &var, const type_list<> &tl,
+                                Args &&...args) {
+  return var;
 }
 
-template <typename T>
-EOS makeBilinearRamp(T &&eos, Real r0, Real a, Real b, Real c) {
-  return BilinearRampEOS<T>(std::forward<T>(eos), r0, a, b, c);
+// The ModifyDispatcher function specifies which tag in tag dispatch
+// to use. It's main purpose is to check whether or not U and Mod<U>
+// are in the type list for the variant. If yes, it dispatches to the
+// ModifyHelper<Mod>(yes,...) method, otherwise, the
+// ModifyHelper<Mod>(no,...)  method.
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyDispatcher(const Variant<Ts...> &var,
+                                const type_list<U, Rest...> &tl, Args &&...args) {
+  constexpr bool type_in_variant =
+      (contains_v<U, Ts...>() && contains_v<Mod<U>, Ts...>());
+  return ModifyHelper<Mod>(bool_constant<type_in_variant>(), var, tl,
+                           std::forward<Args>(args)...);
 }
 
-template <typename T>
-EOS applyShiftAndScale(T &&eos, bool scaled, bool shifted, Real scale, Real shift) {
-  if (shifted && scaled) {
-    ShiftedEOS<T> a(std::forward<T>(eos), shift);
-    ScaledEOS<ShiftedEOS<T>> b(std::move(a), scale);
-    return b;
-  }
-  if (shifted) {
-    return ShiftedEOS<T>(std::forward<T>(eos), shift);
-  }
-  if (scaled) {
-    return ScaledEOS<T>(std::forward<T>(eos), scale);
+// The ModifyHelper function recursively walks through all types in the
+// typelist tl. For each type in tl, it checks if the type is (a) in
+// the variant typelist, and (b) currently held by the variant
+// object. If both (a) and (b) are true, it returns a new variant
+// containing Mod<u>. Otherwise, it moves to the next type.
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyHelper(std::true_type, const Variant<Ts...> &var,
+                            const type_list<U, Rest...> &tl, Args &&...args) {
+  bool contains_u = var.template IsType<U>();
+  if (contains_u) {
+    return Variant<Ts...>(Mod<U>(var.template get<U>(), std::forward<Args>(args)...));
+  } else {
```

Comment:
This is related to above, but unfortunately without `if constexpr` here the function style breaks here. Probably compilers will optimize it away and get something like compile-time behavior, but as C++ code this is explicitly a run-time function.

I need to look a bit deeper. If it's intended as a run-time function, then it can't (or at least _shouldn't_) be used in `constexpr` calls. However it's not difficult to emulated `if constexpr` (you already do).